common/transport/amqp/devdoc/receiver_link_requirements.md

```diff
@@ -0,0 +1,104 @@
+# ReceiverLink Requirements
+
+## Overview
+
+The `ReceiverLink` class is internal to the SDK and shall be considered private. It shall not be used directly by clients of the SDK, who should instead rely on the higher-level constructs (`Client` classes provided by `azure-iot-device` and `azure-iothub`).
+
+The `ReceiverLink` class implements a state machine that manages the underlying `amqp10` link object used to receive messages from IoT Hub. It can be attached and detached manually, and will try to attach automatically if not already attached when setting up a message listener or trying to settle an existing message.
+
+## Example usage
+
+```typescript
+import * as amqp10 from 'amqp10';
+import { AmqpMessage } from './ampq_message';
+
+const linkAddress = 'exampleAddress';
+const amqp10Client = new amqp10.AmqpClient(null);
+
+const receiverLink = new ReceiverLink(linkAddress, null, amqp10Client);
+receiverLink.on('error', (err) => {
+  console.error(err.toString());
+});
+
+receiverLink.on('message', (msg) => {
+  console.log(msg.body);
+  receiverLink.accept(msg, (err) => {
+    if (err) {
+      console.error(err.toString());
+    } else {
+      console.log('message successfully settled (accept)');
+    }
+  });
+});
+```
+
+## Public Interface
+
+### constructor(linkAddress: string, linkOptions: any, amqp10Client: amqp10.AmqpClient)
+
+**SRS_NODE_AMQP_RECEIVER_LINK_16_001: [** The `ReceiverLink` internal state machine shall be initialized in the `detached` state. **]**
+
+**SRS_NODE_AMQP_RECEIVER_LINK_16_002: [** The `ReceiverLink` class shall inherit from `EventEmitter`. **]**
+
+**SRS_NODE_AMQP_RECEIVER_LINK_16_003: [** The `ReceiverLink` class shall implement the `AmqpLink` interface. **]**
+
+### attach(callback: (err?: Error) => void): void
+
+**SRS_NODE_AMQP_RECEIVER_LINK_16_004: [** The `attach` method shall use the stored instance of the `amqp10.AmqpClient` object to attach a new link object with the `linkAddress` and `linkOptions` provided when creating the `ReceiverLink` instance. **]**
+
+**SRS_NODE_AMQP_RECEIVER_LINK_16_005: [** If the `amqp10.AmqpClient` emits an `errorReceived` event during the time the link is attached, the `callback` function shall be called with this error. **]**
+
+**SRS_NODE_AMQP_RECEIVER_LINK_16_006: [** The `ReceiverLink` object should subscribe to the `detached` event of the newly created `amqp10` link object. **]**
+
+**SRS_NODE_AMQP_RECEIVER_LINK_16_007: [** The `ReceiverLink` object should subscribe to the `errorReceived` event of the newly created `amqp10` link object. **]**
+
+**SRS_NODE_AMQP_RECEIVER_LINK_16_008: [** If the `amqp10.AmqpClient` fails to create the link the `callback` function shall be called with this error object. **]**
+
+**SRS_NODE_AMQP_RECEIVER_LINK_16_020: [** The `attach` method shall call the `callback` if the link was successfully attached **]**
```

Comment:
actually, it's undefined. I've discovered recently that using null instead of undefined is a little frowned upon: [https://basarat.gitbooks.io/typescript/docs/tips/null.html](https://basarat.gitbooks.io/typescript/docs/tips/null.html)

---
In reply to: [127616011](https://github.com/Azure/azure-iot-sdk-node/pull/55#discussion_r127616011) [](ancestors = 127616011)